mscore/mixer.cpp

```diff
@@ -302,201 +572,337 @@ void Mixer::changeEvent(QEvent *event)
-      }
+
+
+//MARK:- manage the mixer tree
 
 //---------------------------------------------------------
-//   setPlaybackScore
+//   updateTracks
 //---------------------------------------------------------
 
-void Mixer::setPlaybackScore(Score* score)
+void Mixer::updateTracks()
       {
-      if (_score != score) {
-            _score = score;
-            mixerDetails->setTrack(0);
+      qDebug()<<"Mixer::updateTracks()";
+      const QSignalBlocker blockTreeWidgetSignals(mixerTreeWidget);     // block during this method
+
+      mixerTreeWidget->clear();
+      trackList.clear();
+
+      if (!_score) {
+            disableMixer();
+            return;
+            }
+
+      for (Part* localPart : _score->parts()) {
+            Part* part = localPart->masterPart();
+            //qDebug()<<"Mixer::updateTracks() - outer loop - part->name()"<<part->longName();
+
+            MixerTrackItem* mixerTrackItem = mixerTrackItemFromPart(part);
+
+            QTreeWidgetItem* item = new QTreeWidgetItem(mixerTreeWidget);
+            item->setText(0, part->partName());
+            item->setToolTip(0, part->partName());
+            mixerTreeWidget->addTopLevelItem(item);
+
+            MixerTrackChannel* mixerTrackWidget = new MixerTrackChannel(item, mixerTrackItem);
+            mixerTreeWidget->setItemWidget(item, 1, mixerTrackWidget);
+
+            //Add per channel tracks
+            const InstrumentList* partInstrumentList = part->instruments();
+
+            // Note sure I understand the code here - will this loop ever iterate more
+            // than once?
+            for (auto partInstrumentListItem = partInstrumentList->begin(); partInstrumentListItem != partInstrumentList->end(); ++partInstrumentListItem) {
+
+                  Instrument* instrument = partInstrumentListItem->second;
+                  //qDebug()<<"    Mixer::updateTracks() - inner loop - instrument->trackName()"<<instrument->trackName();
+
+                  if (instrument->channel().size() <= 1)
+                        continue;
+
+                  // add an item for each channel used by a given instrument
+                  for (int i = 0; i < instrument->channel().size(); ++i) {
+                        Channel* channel = instrument->playbackChannel(i, _score->masterScore());
+                        //qDebug()<<"        Mixer::updateTracks() - inner loop, inner loop - channel->name()()"<<channel->name();
+
+
+                        QTreeWidgetItem* child = new QTreeWidgetItem(0);
+                        child->setText(0, channel->name());
+                        child->setToolTip(0, QString("%1 - %2").arg(part->partName()).arg(channel->name()));
+                        item->addChild(child);
+
+                        MixerTrackItem* mixerTrackItem = new MixerTrackItem(MixerTrackItem::TrackType::CHANNEL, part, instrument, channel);
+                        mixerTreeWidget->setItemWidget(child, 1, new MixerTrackChannel(child, mixerTrackItem));
```

Comment:
Looking at Qt documentation on [`setItemWidget()` function](https://doc.qt.io/qt-5/qtreewidget.html#setItemWidget) I found the following paragraph:
> This function should only be used to display static content in the place of a tree widget item. If you want to display custom dynamic content or implement a custom editor widget, use QTreeView and subclass QStyledItemDelegate instead.

`MixerTrackChannel` is apparently intended to be an editor for some channel properties. Did you meet any issues with that approach (perhaps not updating mixer controls if the model gets updated or something like that), and would it be better to use the approach with item delegate instead, like suggested in the documentation?

Reply:
Still the delegates approach looks like something a bit different at first glance... Anyway, it would be good to know your opinion on that!

Reply:
I was put off by that sentence in the docs too, but I don't think it means editing is impossible, just that you have to handle editing yourself. Basically `setItemWidget()` replaces one tree cell with a widget in the view, but it does nothing to the content of that cell in the model. The set widget will ignore model data by default, but it is a real widget for which you can call getter and setter methods to make it reflect (and *affect*) the model, or even display other data entirely that has nothing to do with the model.

No doubt an item delegate would be better though as it would handle the getting and setting for you. However, it does mean learning how to work with models, views, and delegates rather than having them all bundled into a handy widget. Still, if you stick to [QStandardItemModel](https://doc.qt.io/qt-5/qstandarditemmodel.html#details) then it is basically the same as using a QTreeWidget, except that now each cell is an "item" rather than each row.

Reply:
@dmitrio95 Yes, what @shoogle says really. There will be cases where using QStyledEditDelegateItem is going to bring benefits, because the framework has already provided some of the work of connecting the model to what is displayed in the QTreeView. But the docs guidance over-states things. The QStyledEditDelegateItem is just one means of hooking up model items to rows in the tree.

In the previous and current (this PR) mixer code, that hooking up work is being done with signal and slots (provided by QT) and with a MuseScore provided class ChannelListener. And that's plenty. 

The MVC / QStyledEditDelegateItem approach might not fit our case very well because:

*  it's based on the idea of placing an editing widget "on top of the item view while editing takes place". That's not going to helpful here.

* it doesn't, out of the box, provide a slider, (see the QItemEditorFactory docs) - so we'd need to write slider binding code anyway.

* and for us to really and truly get the benefits, I *think* the instruments and parts model MuseScore is currently using would need to be changed to be concrete subclass of QAbstractItemModel. (It makes little sense to create a model just for the mixer when the QTreeWidget provided model is already up to the job.) Making the change elsewhere in MuseScore would, just the purposes of using the QTreeViewWidget in the mixer, be overkill. (That's not to say it might not, at some stage, be worth looking at QAbstractItemModel because it might also bring gains outwith the mixer case. It's more that, that would be part of a bigger project I think.)

Reply:
@shoogle, @Obliquely, thanks for your explanations! If one takes care to manage all connections of data and those widgets manually than all indeed should work well. The only other concern could be that if there are really many items in a model then this widgets-based approach may perform poorly, but that is certainly not the case of a mixer.

Reply:
Model is definitely the better approach long-term, but I agree that you should stick to using a QTreeWidget for now as it is *much easier* to get to grips with. My GSoC student and I have decided to switch to using widgets rather than views for the palettes for this reason, as it was not certain that he would be able to finish the project in time otherwise. If we want to switch back using a separate model later then it should be easier to do this if there is already a widget in place than it would be to start from the custom implementation currently used.

However, I just thought I'd clear up some possible misconceptions about Qt's model-view framework. (I say *possible* misconceptions because I'm not totally sure about any of this because the documentation is contradictory/ambiguous and therefore I haven't actually had the time or courage to try it for myself.)

> [QStyledEditDelegateItem approach is] based on the idea of placing an editing widget "on top of the item view while editing takes place"

The implication is that the widget is somehow hidden or invisible if editing is not happening, but this appears to be another example of poor phrasing in the Qt documentation. [Elsewhere it says](https://doc.qt.io/qt-5/qabstractitemview.html#setItemDelegate) that item delegates handle drawing as well as editing. My guess is that the delegate is just a "picture of a slider" until editing takes place, at which point it is replaced with a real widget.

> it doesn't, out of the box, provide a slider

This is good to know. However, you can [still set a widget](https://doc.qt.io/qt-5/qabstractitemview.html#setIndexWidget) even on an ordinary QTreeView. Of course you don't get the delegate benefits that way, but it might be possible to subclass QTreeView and/or QSlider to make the model update automatic, so it effectively becomes a delegate. I guess that is what you meant by *"we'd need to write slider binding code anyway"*.

> It makes little sense to create a model just for the mixer when the QTreeWidget provided model is already up to the job

It would make it easier to replace it with a proper model later if it wasn't already connected to the view.

Anyway, let's stick with QTreeWidget for now because we *know* it works, whereas the QTreeView is all *"should work"* and *"could work"*.